packages/flutter_tools/lib/src/template.dart

```diff
@@ -0,0 +1,132 @@
+
+      if (relativePath.contains(_kTemplateExtension)) {
+        // If '.tmpl' appears anywhere within the path of this entity, it is
+        // is a candidate for rendering. This catches cases where the folder
+        // itself is a template.
+        _templateFilePaths[relativePath] = path.absolute(entity.path);
+      }
+    }
+  }
+
+  Map<String /* relative */, String /* absolute source */> _templateFilePaths;
+
+  void render(Directory destination, Map<String, dynamic> context,
+      { bool overwriteExisting: true }) {
+    destination.createSync(recursive: true);
+
+    String destinationDirPath = destination.absolute.path;
+
+    _templateFilePaths.forEach((String relativeDestPath,
+        String absoluteSrcPath) {
+      String finalDestinationPath = path
+          .join(destinationDirPath, relativeDestPath)
+          .replaceAll(_kCopyTemplateExtension, '')
+          .replaceAll(_kTemplateExtension, '');
+      File finalDestinationFile = new File(finalDestinationPath);
+      String relativePathForLogging = relativeDestPath
+          .replaceAll(_kCopyTemplateExtension, '')
+          .replaceAll(_kTemplateExtension, '');
+
+      // Step 1: Check if the file needs to be overwritten.
+
+      if (finalDestinationFile.existsSync()) {
+        if (overwriteExisting) {
+          finalDestinationFile.delete(recursive: true);
+          printStatus('$relativePathForLogging exists. Overwriting.');
+        } else {
+          // The file exists but we cannot overwrite it, move on.
+          printStatus('$relativePathForLogging exists. Skipping.');
+          return;
+        }
+      } else {
+        printStatus('$relativePathForLogging created.');
+      }
+
+      finalDestinationFile.createSync(recursive: true);
+      File sourceFile = new File(absoluteSrcPath);
+
+      // Step 2: If the absolute paths ends with a 'copy.tmpl', this file does
+      //         not need mustache rendering but needs to be directly copied.
+
+      if (sourceFile.path.endsWith(_kCopyTemplateExtension)) {
+        finalDestinationFile.writeAsBytesSync(sourceFile.readAsBytesSync());
+
+        return;
+      }
+
+      // Step 3: If the absolute path ends with a '.tmpl', this file needs
+      //         rendering via mustache.
+
+      if (sourceFile.path.endsWith(_kTemplateExtension)) {
+        String templateContents = sourceFile.readAsStringSync();
+        String renderedContents = mustache.render(templateContents, context);
+
+        finalDestinationFile.writeAsStringSync(renderedContents);
+
+        return;
+      }
+
+      // Step 4: This file does not end in .tmpl but is in a directory that
+      //         does. Directly copy the file to the destination.
+
+      finalDestinationFile.writeAsBytesSync(sourceFile.readAsBytesSync());
+    });
+  }
+}
+
+Directory _templateDirectoryInPackage(String name) {
+  String templatesDir = path.join(ArtifactStore.flutterRoot,
+      'packages', 'flutter_tools', 'lib', 'src', 'templates');
```

Comment:
You might also consider putting the `templates` dir into the top-level of the `flutter_tools` project.